direct/connection.go

```diff
@@ -0,0 +1,161 @@
+package direct
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/substrate-client"
+)
+
+const (
+	PongWait     = 40 * time.Second
+	PingInterval = 20 * time.Second
+)
+
+type InnerConnection struct {
+	twinID   uint32
+	session  string
+	identity substrate.Identity
+	url      string
+}
+
+type Writer chan<- []byte
+
+func (w Writer) Write(data []byte) {
+	w <- data
+}
+
+type Reader <-chan []byte
+
+func (r Reader) Read() []byte {
+	return <-r
+}
+
+func NewConnection(identity substrate.Identity, url string, session string, twinID uint32) InnerConnection {
+	return InnerConnection{
+		twinID:   twinID,
+		identity: identity,
+		url:      url,
+		session:  session,
+	}
+}
+
+func (c *InnerConnection) reader(ctx context.Context, cancel context.CancelFunc, con *websocket.Conn, reader chan []byte) {
+	for {
+		typ, data, err := con.ReadMessage()
+		if err != nil {
+			log.Error().Err(err).Msg("failed to read message")
+			cancel()
+			return
+		}
+
+		if typ != websocket.BinaryMessage {
+			log.Error().Msg("invalid message type received")
+			cancel()
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case reader <- data:
+		}
+	}
+}
+
+func (c *InnerConnection) loop(ctx context.Context, con *websocket.Conn, output, input chan []byte) error {
+	defer con.Close()
+
+	local, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pong := make(chan byte)
+	con.SetPongHandler(func(appData string) error {
+		select {
+		case pong <- 1:
+		default:
+		}
+		return nil
+	})
+
+	go c.reader(local, cancel, con, output)
+
+	lastPong := time.Now()
+	for {
+		select {
+		case <-local.Done():
+			return local.Err()
+		case data := <-input:
+			if err := con.WriteMessage(websocket.BinaryMessage, data); err != nil {
+				return err
+			}
+		case <-pong:
+			lastPong = time.Now()
+		case <-time.After(PingInterval):
+			if err := con.WriteControl(websocket.PingMessage, nil, time.Now().Add(10*time.Second)); err != nil {
+				return err
+			}
+
+			if time.Since(lastPong) > PongWait {
+				return fmt.Errorf("connection stalling")
+			}
+		}
+	}
+}
+
+func (c *InnerConnection) Start(ctx context.Context) (Reader, Writer) {
+	output := make(chan []byte)
+	input := make(chan []byte)
+
+	go func() {
```

Comment:
may be in the first 2 lines of the func should be 
```go 
defer close(input)
defer close(output)
```

this means channels are close properly when ctx is cancelled